File: lua/pe016.lua
-- this computes projecteuler.net problem 016 

function double_string(in_string)
-- walk backwards through the string doubling as we go
  local carry = false
  local out_string = ""
  for j = string.len(in_string), 1, -1 do
    local jchar = string.char(string.byte(in_string, j))
  
    if jchar == "0" then
      if carry then
        out_string = "1" .. out_string
      else
        out_string = "0" .. out_string
      end
      carry = false
    end
    if jchar == "1" then
      if carry then
        out_string = "3" .. out_string
      else
        out_string = "2" .. out_string
      end
      carry = false
    end
    if jchar == "2" then
      if carry then
        out_string = "5" .. out_string
      else
        out_string = "4" .. out_string
      end
      carry = false
    end
    if jchar == "3" then
      if carry then
        out_string = "7" .. out_string
      else
        out_string = "6" .. out_string
      end
      carry = false
    end
    if jchar == "4" then
      if carry then
        out_string = "9" .. out_string
      else
        out_string = "8" .. out_string
      end
      carry = false
    end
    if jchar == "5" then
      if carry then
        out_string = "1" .. out_string
      else
        out_string = "0" .. out_string
      end
      carry = true
    end
    if jchar == "6" then
      if carry then
        out_string = "3" .. out_string
      else
        out_string = "2" .. out_string
      end
      carry = true
    end
    if jchar == "7" then
      if carry then
        out_string = "5" .. out_string
      else
        out_string = "4" .. out_string
      end
      carry = true
    end
    if jchar == "8" then
      if carry then
        out_string = "7" .. out_string
      else
        out_string = "6" .. out_string
      end
      carry = true
    end
    if jchar == "9" then
      if carry then
        out_string = "9" .. out_string
      else
        out_string = "8" .. out_string
      end
      carry = true
    end
  end
  if carry then
    out_string = "1" .. out_string
  end
 return out_string
end

local my_string = "2"
local limit = 1000

local ssum = 0
for i =1, limit - 1 do
  my_string = double_string(my_string)
end

ssum = 0
for i = 1, string.len(my_string) do
  ssum = math.floor(string.char(string.byte(my_string, i))) + ssum
end
print(ssum)

File: c++/pe016.cpp
// This is a solution for http://projecteuler.net/problem=16

#include <cstdio>
#include <string>

const int ASCII = 48;

std::string double_string(std::string);

int main() {
  int max = 1000;

// initialise a string to 2 - we store ASCII because it is easier to print
// we could store the raw numbers for a little better performance

  std::string in_string = "1";

  for (int i = 0; i < max; ++i)
    in_string = double_string(in_string);

  int sum = 0;
  for (std::string::iterator it = in_string.begin();
       it < in_string.end();
       it++)
    sum += *it - ASCII;

  printf("%d\n", sum);
  return(0);
}

std::string double_string(std::string in) {
//  given a string containing an integer this function will double the value
//  the string will be extended if necessary to hold the number

  int first = in.at(0) - ASCII;  // use ASCII in string
  int ra = 0;
  if ((first == 5) || (first == 6) || (first == 7) ||
      (first == 8) || (first == 9)) {
    // extend string by a character
    in.resize(in.size() + 1);
    ra = 1;
  }

  // walk backwards through the string doubling as we go
  int carry = 0;
//  for (std::string::reverse_iterator rit = in.rbegin();
//         rit < in.rend(); rit++ )
//  mmm maybe not this walks backwards and we want to modify different points
//  so doing it explicitly using indexing is easier
  for (int j = in.size(); j > 0; --j) {
    int i = j - 1;
    switch (in.at(i)-ASCII) {
      case 0:
        in.at(i+ra) = 0 + ASCII + carry;
        carry = 0;
        break;
      case 1:
        in.at(i+ra) = 2 + ASCII + carry;
        carry = 0;
        break;
      case 2:
        in.at(i+ra) = 4 + ASCII + carry;
        carry = 0;
        break;
      case 3:
        in.at(i+ra) = 6 + ASCII + carry;
        carry = 0;
        break;
      case 4:
        in.at(i+ra) = 8 + ASCII + carry;
        carry = 0;
        break;
      case 5:
        in.at(i+ra) = 0 + ASCII + carry;
        carry = 1;
        break;
      case 6:
        in.at(i+ra) = 2 + ASCII + carry;
        carry = 1;
        break;
      case 7:
        in.at(i+ra) = 4 + ASCII + carry;
        carry = 1;
        break;
      case 8:
        in.at(i+ra) = 6 + ASCII + carry;
        carry = 1;
        break;
      case 9:
        in.at(i+ra) = 8 + ASCII + carry;
        carry = 1;
        break;
    }
  }
  if (carry) {
    in.at(0) = 0 + ASCII + carry;
    carry = 0;
  }
  return in;
}
